src/data_traits.rs

```diff
@@ -22,92 +23,177 @@ use {
+    ///
+    /// If `Self` provides safe mutable access to array elements, then it
+    /// **must** return `Some(_)`.
+    #[doc(hidden)]
+    fn try_is_unique(&mut self) -> Option<bool>;
+}
+
+/// Array representation trait.
+///
+/// An array representation that can be cloned.
+///
+/// ***Internal trait, see `DataRaw`.***
+pub unsafe trait DataRawClone : DataRaw {
+    #[doc(hidden)]
+    /// Unsafe because, `ptr` must point inside the current storage.
+    unsafe fn clone_with_ptr(&self, ptr: *mut Self::Elem) -> (Self, *mut Self::Elem);
+
+    #[doc(hidden)]
+    unsafe fn clone_from_with_ptr(&mut self, other: &Self, ptr: *mut Self::Elem) -> *mut Self::Elem {
+        let (data, ptr) = other.clone_with_ptr(ptr);
+        *self = data;
+        ptr
+    }
+}
+
+/// Array representation trait.
+///
+/// For an array with elements that can be accessed with safe code.
+///
+/// ***Internal trait, see `DataRaw`.***
+pub unsafe trait Data : DataRaw {
     /// Converts the array to a uniquely owned array, cloning elements if necessary.
     #[doc(hidden)]
     fn into_owned<D>(self_: ArrayBase<Self, D>) -> ArrayBase<OwnedRepr<Self::Elem>, D>
     where
         Self::Elem: Clone,
         D: Dimension;
-
-    private_decl!{}
 }
 
 /// Array representation trait.
 ///
-/// For an array with writable elements.
+/// For an array with writable elements that can be accessed with safe code.
 ///
 /// ***Internal trait, see `Data`.***
-pub unsafe trait DataMut : Data {
+//
+// # For implementers
+//
+// If you implement the `DataMut` trait, you are guaranteeing that the
+// `DataRawMut::try_ensure_unique` implementation always panics or ensures that
+// the data is unique. You are also guaranteeing that `try_is_unique` always
+// returns `Some(_)`.
+pub unsafe trait DataMut : Data + DataRawMut {
+    /// Ensures that the array has unique access to its data.
     #[doc(hidden)]
     #[inline]
-    fn ensure_unique<D>(&mut ArrayBase<Self, D>)
-        where Self: Sized,
-              D: Dimension
-    { }
+    fn ensure_unique<D>(self_: &mut ArrayBase<Self, D>)
+    where Self: Sized,
+          D: Dimension
+    {
+        Self::try_ensure_unique(self_)
+    }
 
+    /// Returns whether the array has unique access to its data.
     #[doc(hidden)]
     #[inline]
     fn is_unique(&mut self) -> bool {
-        true
+        self.try_is_unique().unwrap()
     }
 }
 
 /// Array representation trait.
 ///
-/// An array representation that can be cloned.
+/// An array representation that can be cloned and allows elements to be
+/// accessed with safe code.
 ///
 /// ***Internal trait, see `Data`.***
-pub unsafe trait DataClone : Data {
-    #[doc(hidden)]
-    /// Unsafe because, `ptr` must point inside the current storage.
-    unsafe fn clone_with_ptr(&self, ptr: *mut Self::Elem) -> (Self, *mut Self::Elem);
+#[deprecated(note="use `Data + DataRawClone` instead", since="0.12.1")]
+pub trait DataClone : Data + DataRawClone {}
```

Comment:
Hard to decide, but it seems best to me to still keep `DataClone` under its current name, and let the breaking change instead be, that DataClone does not require `Data`. The code using DataClone on github that I've seen (random sample) mostly uses `DataClone + Data<Elem=X>` as bound anyway.

Ideally, we should be here. These should be internal implementation traits. Users should have better ways of being general over our arrays.

Reply:
I'm curious why you lean towards dropping the `Data` bound from `DataClone` instead of adding a `DataRawClone` trait (and keeping `DataClone: Data + DataRawClone` if desired). The name `DataClone` is a little shorter than `DataRawClone`, which is nice for code that needs to specify a `DataClone`/`DataRawClone` bound, but I don't see any other advantages to this approach. If the issue is primarily the length of the name, we could call the new trait `RawClone` instead of `DataRawClone`.

Fwiw, I'd typically write `DataClone<Elem = X>` instead of `DataClone + Data<Elem = X>`. This does come up more often with `DataMut` (example below), but I could conceivably see it happening with `DataClone` too.

```rust
extern crate ndarray;

use ndarray::prelude::*;
use ndarray::{array, Data, DataMut};

fn foo<S1, S2>(a: &ArrayBase<S1, Ix2>, b: &mut ArrayBase<S2, Ix2>)
where
    S1: Data<Elem = f32>,
    S2: DataMut<Elem = f32>, // I like this better than Data<Elem = f32> + DataMut
{
    *b += a;
}

fn main() {
    let a = array![[1., 2.], [3., 4.]];
    let mut b = array![[5., 6.], [7., 8.]];
    foo(&a, &mut b);
    println!("{}", b);
}
```

I'm against dropping the `Data` bound from `DataClone` for two reasons:

1. The biggest issue is inconsistency with the names of other traits: all the other traits named `Data*` have a `Data` bound, while the `DataRaw*` traits don't.
2. If users don't notice the change in the changelog and update their code, it would lead to safety issues for user code relying on `DataClone` implying `Data`. For example, this could happen if existing user code specifies a `DataClone<Elem = A>` bound on an array and dereferences pointers to elements in the array. The consequences of this could be pretty bad, but this coming up in practice does seem unlikely.

We could resolve reason 1 and maintain consistency of naming by dropping the `Data` bound from `DataMut`, `DataOwned`, and `DataShared` too. Then, we'd have something like this:

```rust
/// Array representation trait.
///
/// For an array that meets the invariants of the `ArrayBase` type. This trait
/// does not imply any ownership or lifetime; pointers to elements in the array
/// may not be safe to dereference.
///
/// ***Note:*** `RawData` is not an extension interface at this point.
/// Traits in Rust can serve many different roles. This trait is public because
/// it is used as a bound on public methods.
pub unsafe trait RawData : Sized {
    /// The array element type.
    type Elem;

    #[doc(hidden)]
    // This method is only used for debugging
    fn _data_slice(&self) -> Option<&[Self::Elem]>;

    private_decl!{}
}

/// Array representation trait.
///
/// For an array with writable elements.
///
/// ***Internal trait, see `RawData`.***
pub unsafe trait DataMut : RawData {
    /// If possible, ensures that the array has unique access to its data.
    ///
    /// If `Self` provides safe mutable access to array elements, then it
    /// **must** panic or ensure that the data is unique.
    #[doc(hidden)]
    fn try_ensure_unique<D>(&mut ArrayBase<Self, D>)
    where Self: Sized,
          D: Dimension;

    /// If possible, returns whether the array has unique access to its data.
    ///
    /// If `Self` provides safe mutable access to array elements, then it
    /// **must** return `Some(_)`.
    #[doc(hidden)]
    fn try_is_unique(&mut self) -> Option<bool>;
}

/// Array representation trait.
///
/// A representation that is a unique or shared owner of its data.
///
/// ***Internal trait, see `Data`.***
pub unsafe trait DataOwned : RawData {}

/// Array representation trait.
///
/// A representation that is a lightweight view.
///
/// ***Internal trait, see `Data`.***
pub unsafe trait DataShared : Clone + RawData + DataClone { }

/// Array representation trait.
///
/// An array representation that can be cloned.
///
/// ***Internal trait, see `RawData`.***
pub unsafe trait DataClone : RawData {
    #[doc(hidden)]
    /// Unsafe because, `ptr` must point inside the current storage.
    unsafe fn clone_with_ptr(&self, ptr: *mut Self::Elem) -> (Self, *mut Self::Elem);

    #[doc(hidden)]
    unsafe fn clone_from_with_ptr(&mut self, other: &Self, ptr: *mut Self::Elem) -> *mut Self::Elem {
        let (data, ptr) = other.clone_with_ptr(ptr);
        *self = data;
        ptr
    }
}

/// Array representation trait.
///
/// For an array with elements that can be accessed with safe code.
///
/// ***Internal trait, see `RawData`.***
//
// # For implementers
//
// If you implement both the `Data` and `DataMut` traits, you are guaranteeing
// that the `DataMut::try_ensure_unique` implementation always panics or
// ensures that the data is unique. You are also guaranteeing that
// `try_is_unique` always returns `Some(_)`.
pub unsafe trait Data : RawData {
    #[doc(hidden)]
    fn new(elements: Vec<Self::Elem>) -> Self
    where
        Self: DataOwned;

    /// Converts the data representation to a shared (copy on write)
    /// representation, without any copying.
    #[doc(hidden)]
    fn into_shared(self) -> OwnedRcRepr<Self::Elem>
    where
        Self: DataOwned;

    /// Converts the array to a uniquely owned array, cloning elements if necessary.
    #[doc(hidden)]
    fn into_owned<D>(self_: ArrayBase<Self, D>) -> ArrayBase<OwnedRepr<Self::Elem>, D>
    where
        Self::Elem: Clone,
        D: Dimension;

    /// Ensures that the array has unique access to its data.
    #[doc(hidden)]
    #[inline]
    fn ensure_unique<D>(self_: &mut ArrayBase<Self, D>)
    where
        Self: DataMut + Sized,
        D: Dimension
    {
        Self::try_ensure_unique(self_)
    }

    /// Returns whether the array has unique access to its data.
    #[doc(hidden)]
    #[inline]
    fn is_unique(&mut self) -> bool
    where
        Self: DataMut,
    {
        self.try_is_unique().unwrap()
    }
}
```

There are two disadvantages to this: First, the safety issue (reason 2 above) is more likely to occur with more important traits like `DataMut` missing the `Data` bound. Second, this requires explicitly writing a `Data` bound in more cases (e.g. `DataMut + Data<Elem = A>` instead of `DataMut<Elem = A>`) which isn't as convenient.

> These should be internal implementation traits. Users should have better ways of being general over our arrays.

Treating them as internal would be fine, but leaving them public doesn't seem too bad. If we treat them as internal, we'd need to add `ArrayMut`, `ArrayOwned`, etc., traits implemented on `ArrayBase` to allow users to operate in a generic way over array types.

Reply:
ok, you're right :)